### DISCUSSION TITLE
Random FetchError

### DISCUSSION THREAD

**Discussion by TaylorFam** [vercel/next.js#42386] [Help]
Currently Experiencing Random Fetch errors for builds with next fetching data from a CMS, Headless E-commerce api or other sources, Sometimes the Builds Go Successfully Sometimes they do not.

'FetchError: invalid json response body at 'SomeUrl': Unexpected token < in JSON at position 6' Can't for the life of me find out how to log in next in the build process to log the json body response somewhere else so i can properly diagnose the issue whether its concurrency or other issues. any help pointing in the right direction on how to get the raw response i'm getting which is sometimes my valid json and sometimes throwing errors.

if i turn everything to ServerSideProps there are zero issues. its only getStaticProps that is causing the issues.

**Comment by icyJoseph** [comment]
Usually these ought to start by understanding that:

```
 Unexpected token < in JSON at position 6
```

Is a problem with your code, 99% of the time. Usually we do a fetch request and immediately, in some fashion, call `.json()`.

What happens is that upon lacking of authorization headers, bad payloads, etc, servers respond with various other forms of content. Most notably, fetch doesn't throw on 4xx errors, like for example axios does.

So a bad request might return actual HTML, and you try to resolve that to JSON.

Why would your request not resolve to a 200 with JSON payload is anyone's guess, with this much info I could guess:

- `getStaticProps` is missing auth headers derived from env vars perhaps?
- `getStaticProps` might be using an API incorrectly, perhaps `fs` without `process.cwd()`?

Since the payload is not json, you cannot:

>  Can't for the life of me find out how to log in next in the build process to log the json body response somewhere else

You have to look at the headers content type, or try to call `res.text()`, when the response status code is not 200, or which ever applies. The `res.ok` flag also signals out of 200-299 range statuses. 

However, it is highly likely that the server is just responding something like:

![06ssj9a](https://user-images.githubusercontent.com/21013447/199668606-9c8eae5c-e055-406a-86d6-40130773f6de.png)

So it might be better to spend time into analysing what is it you are sending to the server in that context.

**Reply by TaylorFam** [reply]
Hey @icyJoseph 

i Appreciate the quick and detailed reply, i do want to clarify i do not believe this to be a problem with next or an error from next explicitly, just a frustration that i am unable ( or most probably unaware of how ) to get the information regarding the actual response that is throwing the fetch error due to invalid json. 

my original thought was that the endpoints being requested at build time for page data are all concurrent ( is this correct?), so building large number of pages would possibly have error generated upstream if too many requests are happening at once though inspecting my database and CMS traffic seems like they are not passing their thresholds so i truly cant tell on that. though this Successfully builds sometimes and other times not with no code changes just redeployment from within vercel. and it seems that sometimes one endpoint fails to get json then in a different deployment that failed that same endpoint was successful and a different endpoint that had previously resolved json is now not. 

I do have correct accept and content type headers set for my requests and the behavior persists through this. I do have auth headers and allow origin headers set where applicable, and it is my understanding that if i didn't this would consistently result in build failures not sometimes at seemingly random  builds yes? 

Would an acceptable way to diagnose the issue be then to append the Requests in my application with a conditional check for if (!res.ok) console.log(res.text()) and during the build process anything that would allow me to diagnose the issue or is it possibly for it to be 200 range responses and still not be valid json in the response?

My presumption is that i'm receiving a 500 error from the server through too much traffic something else, i do not believe the requests to be 4xx requests since they are just fetching simple data from endpoints like a strapi server. That in 95 % of other builds is fetched correctly with no issues. 

Your Suggestion to res.text() is much appreciated and hopefully will help me diagnose shortly. Are there any other alternitives i can do such as looking at documentation for the trace files or some way to look at the trace failure somehow that is fired on the fetch error.

I know when the page builds correctly there is a .json file and html page created for that build page are there anywhere the failed pages data is stored so i can access it after the build ( assuming i am testing locally obviously and not on vercel for this process )

Again Thank you So Much for the help i very much appreciate it!
